src/builder/commands/copy.rs

```diff
@@ -30,9 +32,54 @@ impl Copy {
         .member("path", V::Directory::new().is_absolute(true))
         .member("ignore_regex", V::Scalar::new().default(
             r#"(^|/)\.(git|hg|svn|vagga)($|/)|~$|\.bak$|\.orig$|^#.*#$"#))
+        .member("include_regex", V::Scalar::new().optional())
         .member("owner_uid", V::Numeric::new().min(0).optional())
         .member("owner_gid", V::Numeric::new().min(0).optional())
     }
+
+    fn compile_ignore_regex(&self) -> Result<Option<Regex>, RegexError> {
+        Ok(Some(try!(Regex::new(&self.ignore_regex))))
+    }
+
+    fn compile_include_regex(&self) -> Result<Option<Regex>, RegexError> {
+        match self.include_regex {
+            Some(ref include_regex) => {
+                Ok(Some(try!(Regex::new(include_regex))))
+            },
+            None => Ok(None),
+        }
+    }
+}
+
+fn generate_paths(path: &str,
+    ignore_regex: Option<&Regex>, include_regex: Option<&Regex>)
+    -> Option<Vec<PathBuf>>
+{
+    if ignore_regex.map_or(false, |r| r.is_match(path))
+        || include_regex.map_or(false, |r| !r.is_match(path))
+    {
+        return None;
+    }
+
+    let mut result = vec!();
+    let fpath = PathBuf::from(path);
+    match fpath.parent() {
+        Some(parent) => {
+            let mut cur_dir_path = PathBuf::new();
+            for dir_component in parent.components() {
+                match dir_component {
+                    Component::RootDir => cur_dir_path.push("/"),
+                    Component::Normal(dir_name) => cur_dir_path.push(dir_name),
+                    _ => unreachable!(),
+                }
+                result.push(cur_dir_path.clone());
```

Comment:
Wow! For each path `a/b/c/d` you **always** generate a list of allocated paths `a`, `a/b`, `a/b/c` ? I think it's too wasteful.